text/0000-const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,684 @@
+```
+
+is now allowed, because we can prove
+that everything from the creation of the value to the destruction is const evaluable.
+
+Note that one cannot implement `const Drop` for structs with fields with just a regular `Drop` impl.
+While from a language perspective nothing speaks against that, this would be very surprising for
+users. Additionally it would make `const Drop` pretty useless. This is explained in more detail in
+the following subsection about generic code and `const Drop`.
+
+```rust
+struct Foo;
+impl Drop for Foo { fn drop(&mut self) {} }
+struct Bar(Foo);
+impl const Drop for Bar { fn drop(&mut self) {} } // not ok
+// cannot call with `T == Foo`, because of missing `const Drop` impl
+fn foo<T: Drop>(t: T) {
+    // Let t run out of scope and get dropped.
+    // Would not be ok if `T` is `Bar`,
+    // because the drop glue would drop `Bar`'s `Foo` field after the `Bar::drop` had been called.
+    // This function is therefore not accept by the compiler.
+}
+```
+
+### const Drop in generic code
+
+`Drop` is special in Rust. You don't need to specify `T: Drop`, but `T::drop` will still be called
+if an object of type `T` goes out of scope. This means there's an implicit assumption, that given
+an arbitrary `T`, we might call `T::drop` if `T` has a drop impl. While we can specify
+`T: Drop` to allow calling `T::drop` in a `const fn`, this means we can't pass e.g. `u32` for
+`T`, because `u32` has no `Drop` impl. Even types that definitely need dropping, but have no
+explicit `Drop` impl (like `struct Foo(String);`), cannot be passed if `T` requires a `Drop` bound.
+
+To summarize, there are currently three ways to interact with `Drop`:
+
+* don't mention `Drop` in the parameter bounds (or mention `?Drop`, amounting to the same thing)
+    * can pass any type that fulfills the other bounds, but may never go out of scope
+* mention `?const Drop` in the parameter bounds
+    * can only pass types with explicit (const or not) `Drop` impls,
+      still can't drop any values inside the function
+* mention `Drop` in the parameter bounds
+    * can only pass types with explicit `const Drop` impls (so no `u32`)
+
+To resolve this, the language gets a new marker trait `ConstDrop` which is automatically implemented
+for:
+
+* any `Copy` type
+* arrays and tuples consisting solely of `ConstDrop` types
+
+`ConstDrop` is similar to `Copy`: you can only implement it for your type, if all fields are
+`ConstDrop`. While we could automatically implement `ConstDrop` for arbitrary types consisting only
+of other `ConstDrop` types, this would make adding a `!ConstDrop` field to a type a breaking change.
+
+The body of a const function is allowed to generate drop glue for types that implement `ConstDrop`.
+
+To reduce the confusion between `ConstDrop` and `const Drop` for users,
+`impl const Drop for SomeType` requires a `ConstDrop` impl for `SomeType`, similar to how
+implementing `Copy` requires you to also implement `Clone`.
```

Comment:
How's that similar? `Copy` just has `Clone` as the supertrait. Also the "magic" trait is the one being implemented with `Copy`, but the one being required for `const Drop`.